src/commands/publish/route.rs

```diff
@@ -4,48 +4,62 @@ use cloudflare::framework::HttpApiClientConfig;
 
 use crate::http::{api_client, format_error};
 use crate::settings::global_user::GlobalUser;
-use crate::settings::target::{Route, Target};
+use crate::settings::target::Route;
 
-pub fn publish_route(user: &GlobalUser, target: &Target) -> Result<String, failure::Error> {
-    let route = Route::new(&target)?;
-    if route_exists(user, target, &route)? {
-        Ok(route.pattern)
-    } else {
-        create(user, target, &route)?;
-        Ok(route.pattern)
-    }
-}
+pub fn publish_routes(
+    user: &GlobalUser,
+    routes: Vec<Route>,
+    zone_id: &String,
+) -> Result<Vec<Route>, failure::Error> {
+    // get existing routes
+    let existing_routes = fetch_all(user, zone_id)?;
 
-fn route_exists(user: &GlobalUser, target: &Target, route: &Route) -> Result<bool, failure::Error> {
-    let routes = fetch_all(user, target)?;
+    let mut deployed_routes = Vec::new();
 
-    for remote_route in routes {
-        if remote_route == *route {
-            return Ok(true);
+    for route in routes {
+        if existing_routes.contains(&route) {
+            deployed_routes.push(route);
+        } else {
+            match create(user, zone_id, &route) {
+                Ok(_) => deployed_routes.push(route),
+                Err(e) => failure::bail!(
+                    "An error occurred deploying to route {}: {}",
+                    route.pattern,
+                    e
+                ),
+            }
         }
     }
-    Ok(false)
+
+    // TODO: include id info in Route objects
+    Ok(deployed_routes)
 }
 
-fn fetch_all(user: &GlobalUser, target: &Target) -> Result<Vec<Route>, failure::Error> {
+fn fetch_all(user: &GlobalUser, zone_identifier: &String) -> Result<Vec<Route>, failure::Error> {
     let client = api_client(user, HttpApiClientConfig::default())?;
 
     let routes: Vec<Route> = match client.request(&ListRoutes {
-        zone_identifier: &target.zone_id.as_ref().unwrap(),
+        zone_identifier: zone_identifier,
     }) {
         Ok(success) => success.result.iter().map(|r| Route::from(r)).collect(),
         Err(e) => failure::bail!("{}", format_error(e, None)), // TODO: add suggestion fn
     };
 
+    // TODO: include id info in Route objects
     Ok(routes)
 }
 
-fn create(user: &GlobalUser, target: &Target, route: &Route) -> Result<(), failure::Error> {
+// TODO: merge id info into returned Route object
+fn create(
+    user: &GlobalUser,
+    zone_identifier: &String,
```

Comment:
this is interesting to me! why an `&String` and not an `&str` - and how do you even pass an `&String`?

Reply:
i mean, is the difference between a String and a slice. the String comes from the field on the struct, and i need to use it in two places so i borrow it. 🤷‍♀